### PR TITLE
fix(s2/Card): support press event callbacks on standalone Card

### LINKED ISSUE #9674: `Card` component does not support event callbacks when not wrapped in `CardView`
### Provide a general summary of the issue here

The `Card` component states in its documentation that it supports the following props:
- `onAction`
- `onPress`
- `onPressStart`
- `onPressEnd`
- `onPressChange`
- `onPressUp`

However, the "press" actions do not actually trigger any callbacks when the `Card` component is not wrapped in a `CardView`

### 🤔 Expected Behavior?

The onPress- props should be triggered when clicking on the card, [as described by the docs](https://react-spectrum.adobe.com/Card#card). Either that, or remove the documentation saying it's supported.

### 😯 Current Behavior

Callback is not triggered

### 💁 Possible Solution

_No response_

### 🔦 Context

_No response_

### 🖥️ Steps to Reproduce

```tsx
import { Card, Content, Text } from '@react-spectrum/s2';

const MyComponent = () => {
return (
  <Card onPress={(e) => console.log(e)}>
    <Content>
      <Text>Hi there</Text>
    </Content>
  </Card>
);
}
```

### Version

^1.0.0

### What browsers are you seeing the problem on?

Chrome

### If other, please specify.

_No response_

### What operating system are you using?

MacOS - Tahoe 26.2

### 🧢 Your Company/Team

Adobe/Incubator

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by starboyvarun**
## Summary

Closes #9674

Standalone `Card` components (used outside `CardView`) were silently discarding all press-event props (`onPress`, `onPressStart`, `onPressEnd`, `onPressChange`, `onPressUp`) and `onAction`.  
The root cause was that the non-`CardView`, non-link render path returned a plain `<div>` that never wired up any interaction handling.

### What changed

- When at least one press/action callback is supplied on a standalone `Card`, it now renders as `role="button"` with `tabIndex={0}` and uses three react-aria hooks — `usePress`, `useHover`, and `useFocusRing` — to handle interactions.
- Hover, focus-visible, and pressed states are forwarded into `InternalCardContext` and the `card()` style function, so visual feedback (shadow elevation, focus ring, press scale) works just like inside `CardView`.
- `onAction` is called alongside `onPress` on a successful press, matching the activation semantics users expect from a standalone interactive card.
- `isDisabled` disables all callbacks, removes the card from the tab order, and sets `aria-disabled`.
- Cards with no press props continue to render as a static `<div>` — no behaviour change.
- Cards with `href` continue to render as a `<Link>` — no behaviour change.
- Hooks are called unconditionally to satisfy React's rules of hooks; they are disabled via `isDisabled` when the interactive-standalone path is not active.

### New imports (all already dependencies of the package)
| Import | Source |
|---|---|
| `usePress` | `react-aria/usePress` |
| `useHover` | `react-aria/useHover` |
| `useFocusRing` | `react-aria/useFocusRing` |
| `mergeProps` | `react-aria/mergeProps` |

### Tests added

`packages/@react-spectrum/s2/test/Card.test.tsx` — six cases covering:
- Static card renders without `role` / `tabindex`
- `onPress` fires on click
- `onAction` fires on click
- Both `onPress` and `onAction` fire together
- `onPressStart` / `onPressEnd` fire
- Disabled card: no callbacks fire, `aria-disabled` set, removed from tab order

## Test plan
- [ ] Run `yarn test packages/@react-spectrum/s2/test/Card.test.tsx` — all six new tests pass
- [ ] Verify existing `CardView` stories are unaffected
- [ ] Manual smoke-test: render `<Card onPress={…}>` standalone and confirm callback fires on click and Enter/Space

### CODE REVIEW

**Comment by starboyvarun** [comment]
@devongovett can you please review my PR?

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 1]
```diff
@@ -486,6 +528,48 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
   }
 
   if (ElementType === 'div' || isSkeleton) {
+    if (isInteractiveStandalone) {
+      return (
+        <div
+          {...mergeProps(filterDOMProps(otherProps), pressProps, hoverProps, focusProps)}
+          id={id != null ? String(id) : undefined}
+          ref={domRef}
+          role="button"
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/Card.tsx]: Instead of using usePress and useHover etc and specifying role="button" we should just be able to use `useButton`

it should take care of the tabIndex and aria-disabled as well

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      <Card>
+        <Content>
+          <Text slot="title">Static Card</Text>
+        </Content>
+      </Card>
+    );
+    let el = getByText('Static Card').closest('[class]')!.parentElement!;
+    expect(el.tagName).toBe('DIV');
+    expect(el).not.toHaveAttribute('role');
+    expect(el).not.toHaveAttribute('tabindex');
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx]: could do something like
```suggestion
expect(queryAllByRole('button')).toHaveLength(0);
await user.tab();
expect(document.activeElement).toBe(document.body);
```

then it's less dependent on structure and implementation

note, the above is psuedocode, it may need some tweaks, but that's the idea

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  it('does not expose role=button when no press callbacks are provided', () => {
+    let {queryByRole} = render(
+      <Card>
+        <Content>
+          <Text slot="title">Static Card</Text>
+        </Content>
+      </Card>
+    );
+    expect(queryByRole('button')).toBeNull();
+  });
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:173]:
```suggestion
```

This is a duplicate of the first test

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      </Card>
+    );
+
+    let card = getByRole('button');
+    expect(card).not.toHaveAttribute('tabindex');
+    expect(card).toHaveAttribute('aria-disabled', 'true');
+
+    await user.click(card);
+    expect(onPress).not.toHaveBeenCalled();
+  });
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:131]: We'll want a test using the keyboard. And we should check onHover and onFocus

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 1 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Done — swapped to `useButton({elementType: 'div', ...}, domRef)`, matching the pattern already used by `StepButton` in `NumberField.tsx`. It now supplies `role`, `tabIndex`, and `aria-disabled` itself, so I removed the three manual attributes. Kept `useHover`/`useFocusRing` separate since `useButton` doesn't cover hover/focus-visible visual state.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 2 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx]: Updated to exactly this — `queryAllByRole('button')).toHaveLength(0)` plus `user.tab()` / `document.activeElement === document.body`. Thanks, much less coupled to internal markup than the old `.closest()` traversal.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 3 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx:173]: Removed.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 4 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx:131]: Added a keyboard-activation test (Enter and Space via `user.keyboard`) and a test that hovering/tabbing changes the rendered class. `Card` doesn't expose raw `onHover`/`onFocus` props, so the latter verifies the internal hover/focus-visible state is actually wired into the style output rather than just checking a callback fired.

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2025 Adobe. All rights reserved.
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:2]:
```suggestion
 * Copyright 2026 Adobe. All rights reserved.
```

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+    act(() => jest.runAllTimers());
+  });
+
+  afterAll(() => {
+    jest.restoreAllMocks();
+  });
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:33]:
```suggestion
```

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    );
+    expect(queryAllByRole('button')).toHaveLength(0);
+    await user.tab();
+    expect(document.activeElement).toBe(document.body);
+  });
+
+  it('renders as role=button and fires onPress when onPress is provided', async () => {
+    let onPress = jest.fn();
+    let {getByRole} = render(
+      <Card onPress={onPress}>
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:51]: This does add button role, but there's no guarantee that someone will pass in Content with a title. We'll also need to allow aria-label and aria-labelledby. I don't think we can enforce it, same as S2 button

File: packages/@react-spectrum/s2/test/Card.test.tsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        <Content>
+          <Text slot="title">Both Callbacks Card</Text>
+        </Content>
+      </Card>
+    );
+
+    let card = getByRole('button');
+    await user.click(card);
+    expect(onPress).toHaveBeenCalledTimes(1);
+    expect(onAction).toHaveBeenCalledTimes(1);
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/test/Card.test.tsx:95]: it'd be good to assert the order here as well

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 9]
```diff
@@ -420,10 +426,49 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
...
+      elementType: 'div',
+      onPress: e => {
+        onPress?.(e);
+        onAction?.();
+      },
+      onPressStart,
+      onPressEnd,
+      onPressChange,
+      onPressUp,
+      isDisabled: isDisabled || !isInteractiveStandalone
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/Card.tsx:463]: rather than making it disabled, we should just not spread the buttonProps on the target

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 10]
```diff
@@ -420,10 +426,49 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
...
+    },
+    domRef
+  );
+  let {hoverProps, isHovered: isInteractiveHovered} = useHover({
+    isDisabled: isDisabled || !isInteractiveStandalone
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/Card.tsx:468]: same thing for this and useFocusRing

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 11]
```diff
@@ -420,10 +426,49 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
...
+    isDisabled,
     ...otherProps
   } = props;
   let isQuiet = variant === 'quiet';
   let isSkeleton = useIsSkeleton();
+
+  // True when the card is used standalone (not inside CardView) and the caller
+  // has provided at least one press/action callback.
+  let isInteractiveStandalone =
+    ElementType === 'div' &&
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/Card.tsx:444]: the other ElementType is link?
technically links support press props and all of that, just not onAction usually...
https://react-spectrum.adobe.com/Link#api

Is there a reason why you chose this condition?

maybe some tests as standalone links would be useful? maybe it already works. If that's the base, we could probably swap a lot of this out to make it work the same way but using Button from RAC with the elementType of div so it takes care of the role and tab index etc.

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 12]
```diff
@@ -420,10 +426,49 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
...
+    !isSkeleton &&
+    !props.href &&
+    !!(onPress || onPressStart || onPressEnd || onPressChange || onPressUp || onAction);
+
+  // Hooks must be called unconditionally (React rules of hooks).
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/Card.tsx:449]: This is common info and lint will yell if you try to break it, we can remove this comment

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 5 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx]: Applied — updated to `Copyright 2026`.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 6 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx]: Removed. You're right that it was doing nothing here — `restoreAllMocks` only undoes `jest.spyOn`/`jest.mock`, and this file only uses plain `jest.fn()` callbacks, which the `jest.clearAllMocks()` in `afterEach` already covers. I also checked the sibling S2 test files and none of them carry this block, so it was just boilerplate that made this file look different for no reason.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 7 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx:47]: Good catch, and this was a real hole — thanks. There were actually two things broken here:

1. `CardProps` never declared `aria-label` / `aria-labelledby`, so passing them wasn't even type-legal.
2. Even if a consumer forced them through, `filterDOMProps` **silently strips** the labelling props unless you opt in with `{labelable: true}` — so they'd never have reached the DOM.

Fixed both: `CardProps` now extends `AriaLabelingProps` from `@react-types/shared` (mirroring how RAC's `Button` picks it up via `AriaButtonProps` → `AriaBaseButtonProps`), and all three `filterDOMProps` call sites in `Card` now pass `{labelable: true}`. Added a test that renders an interactive `Card` whose `Content` has no `title` slot and asserts it's still reachable as `getByRole('button', {name: 'Labelled card'})`.

Agreed on not enforcing it — I've matched S2 `Button`'s stance and left it as an escape hatch rather than adding a dev-time warning.

One question, since you'd know the intended direction better than I would: for the `CardView` path the item gets its accessible name from the card contents via `GridListItem`'s `textValue`/content. Would you want a standalone interactive `Card` to fall back to something similar (e.g. deriving a name from the `title` slot when no `aria-label` is given), or is relying on the content's own text — exactly as it works today — the behaviour you'd expect? Happy to add the fallback if you think it's worth the extra machinery; I left it out for now to keep this PR focused on the press-props fix.

File: packages/@react-spectrum/s2/test/Card.test.tsx (hunk 8 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/test/Card.test.tsx:91]: Added — the test now pins the ordering with `invocationCallOrder`, asserting `onPress` fires before `onAction`:

```js
expect(onPress.mock.invocationCallOrder[0]).toBeLessThan(onAction.mock.invocationCallOrder[0]);
```

Worth pinning since the order is a real contract of the implementation (both are dispatched from the single `onPress` handler passed to `useButton`), and nothing would have caught a regression if they were swapped.

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 9 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Fixed — `useButton` now receives the real `isDisabled` prop:

```js
let {buttonProps, isPressed: isInteractivePressed} = useButton(
  {elementType: 'div', onPress: …, onPressStart, onPressEnd, onPressChange, onPressUp, isDisabled},
  domRef
);
```

You're right that the gating was in the wrong place. The component already branches on `isInteractiveStandalone`, and the non-interactive branches simply never spread `buttonProps`, so the `|| !isInteractiveStandalone` was expressing the same condition a second time — and doing it by telling the hook the card was *disabled*, which isn't what "rendering as a static div" means. Now `isDisabled` says exactly what it says, and whether the props reach the DOM is decided solely by which branch renders.

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 10 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Done — `useHover` gets the same treatment and is now just:

```js
let {hoverProps, isHovered: isInteractiveHovered} = useHover({isDisabled});
```

`useFocusRing` was already called with no arguments (it has no `isDisabled` option), and like the others its `focusProps` are only spread in the interactive-standalone branch, so nothing was needed there.

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 12 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Removed. Agreed — it restated a universal React rule that lint already enforces, so it was just noise. The second half of it described the `isDisabled: … || !isInteractiveStandalone` trick, which is gone now anyway per your other comment.

File: packages/@react-spectrum/s2/src/Card.tsx [hunk 13]
```diff
@@ -488,9 +533,59 @@ export const Card = forwardRef(function Card(props: CardProps, ref: DOMRef<HTMLD
   }
 
   if (ElementType === 'div' || isSkeleton) {
+    if (isInteractiveStandalone) {
```
Comment (devongovett) [inline packages/@react-spectrum/s2/src/Card.tsx]: Getting to be a bit complex. Lots shared with the Link case, perhaps we can clean it up a bit more.

**Review by devongovett — DISMISSED** [review]

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 13 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Agreed — done in d97ce9652.

After the switch to RAC `Button`, the link branch and the button branch were rendering the same thing twice: identical `onPress`/`onPressStart`/`onPressEnd`/`onPressChange`/`onPressUp`/`isDisabled` forwarding, identical `className` and `style` callbacks, and an identical `InternalCardContext.Provider` body. The only real difference was which element got rendered.

Since `Link` and `Button` are both RAC render prop components exposing the same interaction state, they can share all of it. The two branches are now one block that defines `pressProps`, `getClassName`, `getStyle`, and `renderCard` once and only picks the element at the very end:

```jsx
return isLink ? (
  <Link {...filterDOMProps(otherProps, {isLink: true, labelable: true})} {...pressProps} ref={domRef as any} className={getClassName} style={getStyle}>
    {renderCard}
  </Link>
) : (
  <RACButton {...filterDOMProps(otherProps, {labelable: true})} {...pressProps} ref={domRef as unknown as RefObject<HTMLButtonElement>} className={getClassName} style={getStyle}>
    {renderCard}
  </RACButton>
);
```

The shared callbacks are typed against a small `CardInteractionState` (`isHovered`/`isFocusVisible`/`isPressed`/`isDisabled`), which both `LinkRenderProps` and `ButtonRenderProps` are supersets of, so this stays fully typed with no casts beyond the two refs that were already there.

Net effect is 86 lines removed for 94 added (the additions are mostly the shared definitions and one new test), and the merged `onPress`/`onAction` handler now exists in exactly one place, so activation semantics can't drift between the link and button paths.

File: packages/@react-spectrum/s2/src/Card.tsx (hunk 11 above)
Comment (starboyvarun) [reply packages/@react-spectrum/s2/src/Card.tsx]: Thanks for pushing the RAC `Button` commit — that answered the second half of this better than I would have. Answering the first half, and I've now folded the two paths together in d97ce9652.

**On the condition.** `ElementType` here isn't a link/div choice — it comes from `InternalCardViewContext` and is only ever `'div'` (standalone) or `GridListItem` (inside a `CardView`):

```js
export const InternalCardViewContext = createContext({
  ElementType: 'div' as 'div' | typeof GridListItem,
  layout: 'grid' as 'grid' | 'waterfall'
});
```

So `ElementType === 'div'` just means "standalone", and link-ness comes from `props.href`, which the branch above already handled by returning a `<Link>`. The `!props.href` in my condition was therefore only saying "the link branch didn't already claim this render", not "links can't be pressed". It read as a policy decision when it was really just branch bookkeeping — which is why the question was fair.

**On links supporting press props — you're right, and they already do.** Press props are forwarded on the link path too (that landed a couple of commits before your review), so `onPress`/`onPressStart`/etc. work on an `href` card. `onAction` fires there as well; I kept it consistent rather than making it button-only, since a consumer passing `onAction` to a link card almost certainly means "also do this on activation" rather than "silently ignore me". Happy to drop it from the link path if you'd rather keep `onAction` strictly a CardView/button concept.

**Tests for standalone links** — there are now three:
- `forwards press callbacks to the underlying link when href is provided` (click fires both `onPress` and `onAction`)
- `fires press callbacks on a link card activated with the keyboard` (Enter — new in this commit, and it exercises the now-shared path)
- `does not fire press callbacks on a disabled link card` (also asserts it's out of the tab order)

All use fragment hrefs (`href="#link-card"`), since jsdom throws `Not implemented: navigation` on a real href and `setupTests.js` rethrows `console.error` as a failure — which only shows up on the React 16/17 jobs.

**On collapsing the two.** With `Button` and `Link` both being RAC render prop components over the same interaction state, everything except the element itself is now shared — see my reply on Devon's thread for the shape. The condition is also no longer doing bookkeeping; it reads as `if (isLink || isInteractiveStandalone)`.

Verified green on React 19, 17.0.2, and 16.14 (`@testing-library/react@12`) — 12 tests, plus `tsgo --noEmit` and `oxlint` clean.

**Review by snowystinger — APPROVED** [review]
